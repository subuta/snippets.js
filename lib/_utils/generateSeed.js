import {validateModel} from 'lib/_utils/ModelConfig'
import _ from 'lodash'
import faker from 'faker'
import fakerEn from 'faker/locale/en_US'
import { snippets as s } from 'bld.js'

import hash from 'string-hash'

export default (modelConfig, locale = 'en') => {
  validateModel(modelConfig)

  // use hash of json as a randomness
  faker.seed(hash(JSON.stringify(modelConfig)));

  // set locale.
  faker.locale = locale

  let {tableName, properties} = modelConfig.schema
  tableName = s.tableName(tableName)

  return _.transform(properties, (result, prop, key) => {
    result[key] = (function generator (prop, key) {
      const {type} = prop
      const _key = _.snakeCase(key)

      // generate data for well-known column.
      if (_key === 'zip_code') {
        return faker.address.zipCode()
      }

      if (type === 'string' && _.includes(_key, 'address')) {
        return faker.address.streetAddress()
      }

      if (_.includes(_key, 'city')) {
        return faker.address.city()
      }

      if (_.endsWith(_key, 'price')) {
        return faker.commerce.price()
      }

      if (_.includes(_key, 'state')) {
        return faker.address.state()
      }

      if (_.includes(_key, 'country_code')) {
        return faker.address.countryCode()
      }

      if (_.includes(_key, 'country')) {
        return faker.address.country()
      }

      if (_key === 'last_name') {
        return faker.name.lastName()
      }

      if (_.includes(_key, 'color')) {
        return faker.commerce.color()
      }

      if (_key === 'first_name') {
        return faker.name.firstName()
      }

      if (_.includes(_key, 'phone') || _.startsWith(_key, 'tel')) {
        return faker.phone.phoneNumber()
      }

      if (_.includes(_key, 'mail')) {
        return fakerEn.internet.email()
      }

      if (_.includes(_key, 'tag') || _.includes(s.singular(_key), 'category')) {
        return [faker.random.word()]
      }

      if (_.startsWith(_key, 'last') && _.endsWith(_key, 'at')) {
        return faker.date.past()
      }

      if (key.match(/starts?_at/)) {
        return faker.date.past()
      }

      if (key.match(/ends?_at/)) {
        return faker.date.future()
      }

      if (_.endsWith(_key, 'at') || _.endsWith(_key, 'date')) {
        return faker.date.future()
      }

      if (_key === 'uuid') {
        return faker.random.uuid()
      }

      if (_.endsWith(_key, 'url')) {
        return faker.internet.url()
      }

      if (_key === 'id' || _.endsWith(_key, 'id')) {
        return faker.random.number()
      }

      if (_.endsWith(_key, 'count') || _.endsWith(_key, 'counts')) {
        return faker.random.number()
      }

      if (_key === 'locale') {
        return faker.random.locale()
      }

      if (_.includes(_key, 'avatar')) {
        return faker.internet.avatar()
      }

      if (_.includes(_key, 'image')) {
        return faker.image.image()
      }

      if (_.includes(tableName, 'attachment') && (_key === 'url' || _key === 'image')) {
        return faker.image.cats()
      }

      if (_.includes(tableName, 'companies') && (_key === 'name' || _key === 'label')) {
        return faker.company.companyName()
      }

      if (_.includes(tableName, 'user') && (_key === 'name')) {
        if (locale === 'ja') return `${faker.name.lastName()} ${faker.name.firstName()}`

        return faker.name.findName()
      }

      if ((_.includes(tableName, 'product') || _.includes(tableName, 'item')) && _key === 'name') {
        return faker.commerce.productName()
      }

      if (_key === 'user_name' || _key === 'username') {
        return fakerEn.internet.userName()
      }

      // comment
      if (_key === 'comment' || (_.includes(tableName, 'comment') && _key === 'text')) {
        return fakerEn.lorem.paragraph()
      }

      if (_key === 'content' || _key === 'body') {
        return fakerEn.lorem.paragraph()
      }

      // string specific conditions
      if (type === 'string' && prop.format) {
        switch (prop.format) {
          case 'date':
          case 'date-time':
            return faker.date.future().toISOString()
          case 'uri':
            return faker.internet.url()
          case 'email':
            return fakerEn.internet.email()
          case 'ipv4':
            return faker.internet.ip()
          case 'ipv6':
            return faker.internet.ipv6()
        }
      }

      // then try to generate by column-type.
      switch (type) {
        case 'array':
          return [generator(prop.items, key)]
        case 'object':
          return _.transform(prop.properties || {}, (result, p, k) => {
            result[k] = generator(p, k)
          }, {})
        case 'number':
        case 'integer':
          return faker.random.number()
        case 'boolean':
          return faker.random.boolean()
        default:
          return faker.random.words()
      }
    })(prop, key)
  }, {})
}
