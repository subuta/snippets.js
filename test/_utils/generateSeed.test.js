/* global expect, describe, it, beforeEach, afterEach */

import _ from 'lodash'

import { build, format, snippets as s } from 'bld.js'

import generateSeed, { generateSeeds } from 'lib/_utils/generateSeed'
import sinon from 'sinon'

describe('utils/generateSeed', () => {
  let clock

  beforeEach(() => {
    clock = sinon.useFakeTimers()
  })

  afterEach(() => {
    clock.restore()
  })

  it('should return seed data for fake user', () => {
    const seed = generateSeed({
      schema: {
        tableName: 'users',
        properties: {
          name: {
            type: 'string'
          },

          userName: {
            type: 'string'
          },

          loggedInCount: {
            type: 'number'
          },

          email: {
            type: 'string'
          },

          avatar: {
            type: 'string'
          },

          locale: {
            type: 'string'
          },

          lastSignInAt: {
            type: 'string',
            format: 'date'
          }
        }
      }
    }, 0, 'ja')

    expect(seed).toMatchSnapshot()
  })

  it('should return seed data for fake book', () => {
    const seed = generateSeed({
      schema: {
        tableName: 'books',
        properties: {
          title: {
            type: 'string'
          },

          authorId: {
            type: 'number'
          },

          price: {
            type: 'string'
          },

          dateStr: {
            type: 'string',
            format: 'date'
          }
        }
      }
    })

    expect(seed).toMatchSnapshot()
  })

  it('should return data for array with items', () => {
    const seed = generateSeed({
      schema: {
        tableName: 'article',
        properties: {
          tag: {
            type: 'array',
            items: {type: 'string'}
          }
        }
      }
    })

    expect(seed).toMatchSnapshot()
  })

  it('should return data for object with items', () => {
    const seed = generateSeed({
      schema: {
        tableName: 'users',
        properties: {
          address: {
            type: 'object',
            properties: {
              'streetAddress': {'type': 'string'},
              'city': {'type': 'string'},
              'state': {'type': 'string'}
            }
          },

          countryCode: {
            type: 'string'
          }
        }
      }
    })

    expect(seed).toMatchSnapshot()
  })

  it('should return data for company name', () => {
    const seed = generateSeed({
      schema: {
        tableName: 'companies',
        properties: {
          name: {
            type: 'string'
          }
        }
      }
    })

    expect(seed).toMatchSnapshot()
  })
})

describe('utils/generateSeeds', () => {
  let clock

  beforeEach(() => {
    clock = sinon.useFakeTimers()
  })

  afterEach(() => {
    clock.restore()
  })

  it('should return seed data for fake 3 user', () => {
    const seed = generateSeeds({
      schema: {
        tableName: 'users',
        properties: {
          name: {
            type: 'string'
          },

          userName: {
            type: 'string'
          },

          loggedInCount: {
            type: 'number'
          },

          email: {
            type: 'string'
          },

          avatar: {
            type: 'string'
          },

          locale: {
            type: 'string'
          },

          lastSignInAt: {
            type: 'string',
            format: 'date'
          }
        }
      }
    }, 3, 'ja')

    expect(seed).toMatchSnapshot()
  })
})
