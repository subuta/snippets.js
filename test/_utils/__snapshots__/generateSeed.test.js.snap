// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/generateJunctionSeed should return single seed data for junction-table model 1`] = `
Object {
  "articleId": 52697,
  "id": 48239,
  "tagId": 82337,
}
`;

exports[`utils/generateJunctionSeeds should return multiple seed data for junction-table model 1`] = `
Array [
  Object {
    "articleId": 52697,
    "id": 48239,
    "tagId": 82337,
  },
  Object {
    "articleId": 14400,
    "id": 52023,
    "tagId": 43149,
  },
  Object {
    "articleId": 29655,
    "id": 37083,
    "tagId": 3838,
  },
]
`;

exports[`utils/generateSeed should return signle data for array with items 1`] = `
Object {
  "id": 50444,
  "tag": Array [
    "South Dakota",
  ],
}
`;

exports[`utils/generateSeed should return single data for company name 1`] = `
Object {
  "name": "Wolff and Sons",
}
`;

exports[`utils/generateSeed should return single data for object with items 1`] = `
Object {
  "address": Object {
    "city": "Port Liza",
    "state": "South Carolina",
    "streetAddress": "202 Johns Stream",
  },
  "countryCode": "KP",
  "id": 28085,
}
`;

exports[`utils/generateSeed should return single seed data for fake book 1`] = `
Object {
  "authorId": 18129,
  "dateStr": "2000-02-22T08:50:18.260Z",
  "id": 78061,
  "price": "472.00",
  "slug": "quam-labore-voluptatum",
  "title": "blue Home Loan Account Bedfordshire",
}
`;

exports[`utils/generateSeed should return single seed data for fake user 1`] = `
Object {
  "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/uxalex/128.jpg",
  "email": "Hilbert_Rath@gmail.com",
  "id": 41704,
  "lastSignInAt": 1999-10-11T11:21:53.655Z,
  "locale": "zh_CN",
  "loggedInCount": 16436,
  "name": "松本 結愛",
  "userName": "Laury24",
}
`;

exports[`utils/generateSeed should return single seed data for fake user with today(1970/1/1) 1`] = `
Object {
  "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/uxalex/128.jpg",
  "email": "Hilbert_Rath@gmail.com",
  "id": 41704,
  "lastSignInAt": 1969-10-11T01:21:53.655Z,
  "locale": "zh_CN",
  "loggedInCount": 16436,
  "name": "Guadalupe Parker",
  "userName": "Laury24",
}
`;

exports[`utils/generateSeed should return single seed data for fake user without locale 1`] = `
Object {
  "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/uxalex/128.jpg",
  "email": "Hilbert_Rath@gmail.com",
  "id": 41704,
  "lastSignInAt": 1999-10-11T11:21:53.655Z,
  "locale": "zh_CN",
  "loggedInCount": 16436,
  "name": "Guadalupe Parker",
  "userName": "Laury24",
}
`;

exports[`utils/generateSeedWithRelation should return single seed data with relation 1`] = `
Object {
  "attachmentId": "28d15c5a-a70c-48e4-9772-bc910f421907",
  "channelId": 68264,
  "commentedById": 54551,
  "id": 57447,
  "text": "Ut amet molestiae aliquam numquam ad ut et. Et ut perspiciatis fugiat expedita fuga hic odio. Ducimus cumque aut harum in perferendis veritatis tempora aut. Facilis officia sed. Eum excepturi officia ipsum sunt voluptatem eius. Rerum qui doloremque ex velit porro animi odio saepe sunt.",
}
`;

exports[`utils/generateSeedWithRelation should return single seed data with relation as instance if asId = false 1`] = `
Object {
  "attachment": Object {
    "id": "28d15c5a-a70c-48e4-9772-bc910f421907",
    "imageUrl": "http://lorempixel.com/640/480/nature",
    "name": "one-to-one",
    "type": "embrace",
  },
  "channel": Object {
    "id": 68264,
    "name": "convergence Administrator Clothing",
  },
  "commentedBy": Object {
    "auth0Id": "4be76d72-37c6-4af4-9b6d-5cf3cb62d63a",
    "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/devankoshal/128.jpg",
    "id": 54551,
    "locale": "en_au_ocker",
    "nickname": "Juana5",
    "status": "New Israeli Sheqel",
  },
  "id": 57447,
  "text": "Ut amet molestiae aliquam numquam ad ut et. Et ut perspiciatis fugiat expedita fuga hic odio. Ducimus cumque aut harum in perferendis veritatis tempora aut. Facilis officia sed. Eum excepturi officia ipsum sunt voluptatem eius. Rerum qui doloremque ex velit porro animi odio saepe sunt.",
}
`;

exports[`utils/generateSeeds should return multiple seed data for fake 3 user 1`] = `
Array [
  Object {
    "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/uxalex/128.jpg",
    "email": "Hilbert_Rath@gmail.com",
    "id": 41704,
    "lastSignInAt": 1999-10-11T11:21:53.655Z,
    "locale": "zh_CN",
    "loggedInCount": 16436,
    "name": "松本 結愛",
    "userName": "Laury24",
  },
  Object {
    "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/ChrisFarina78/128.jpg",
    "email": "Lila_Bauch@hotmail.com",
    "id": 33681,
    "lastSignInAt": 1999-12-10T02:02:47.036Z,
    "locale": "en",
    "loggedInCount": 64062,
    "name": "山田 陽翔",
    "userName": "Hilma_Lesch",
  },
  Object {
    "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/arindam_/128.jpg",
    "email": "Sonny_Luettgen25@hotmail.com",
    "id": 86676,
    "lastSignInAt": 1999-12-29T05:55:54.455Z,
    "locale": "sk",
    "loggedInCount": 11531,
    "name": "林 陸斗",
    "userName": "Dax65",
  },
]
`;

exports[`utils/generateSeedsWithRelation should return multiple seed data with relation 1`] = `
Array [
  Object {
    "attachmentId": "28d15c5a-a70c-48e4-9772-bc910f421907",
    "channelId": 68264,
    "commentedById": 54551,
    "id": 57447,
    "text": "Ut amet molestiae aliquam numquam ad ut et. Et ut perspiciatis fugiat expedita fuga hic odio. Ducimus cumque aut harum in perferendis veritatis tempora aut. Facilis officia sed. Eum excepturi officia ipsum sunt voluptatem eius. Rerum qui doloremque ex velit porro animi odio saepe sunt.",
  },
  Object {
    "attachmentId": "abfbc020-04b2-4d70-a1ed-aa85e702e181",
    "channelId": 36069,
    "commentedById": 15599,
    "id": 33984,
    "text": "Id optio rem culpa minus. Nostrum aut et. Aliquam repellat est reprehenderit.",
  },
  Object {
    "attachmentId": "62fc4caf-224d-4658-949e-b68307f974d2",
    "channelId": 63669,
    "commentedById": 34490,
    "id": 46230,
    "text": "Facilis cupiditate quo nemo hic. Rem nesciunt aliquid aut qui architecto saepe enim laborum. Quia dolor consequuntur et dolorem tenetur totam nisi asperiores.",
  },
]
`;
